src/log_batch.rs

```diff
@@ -392,208 +332,363 @@ impl<M: MessageExt> LogItem<M> {
-        if batch_len > buf.len() {
-            return Err(Error::TooShort);
+        for item in items.items.iter_mut() {
+            if let LogItemContent::EntriesIndex(entries_index) = &mut item.content {
+                for ei in entries_index.0.iter_mut() {
+                    ei.compression_type = compression_type;
+                    ei.entries_offset = entries_offset;
+                    ei.entries_len = entries_len;
+                }
+            }
         }
-        let batch_type = CompressionType::from_byte(header as u8);
-        test_batch_checksum(&buf[..batch_len])?;
-        trace!("decoding log batch({}, {:?})", batch_len, batch_type);
+        buf.consume(CHECKSUM_LEN);
 
-        let decompressed = match batch_type {
-            CompressionType::None => Cow::Borrowed(&buf[..(batch_len - CHECKSUM_LEN)]),
-            CompressionType::Lz4 => Cow::Owned(decompress(&buf[..batch_len - CHECKSUM_LEN])),
-        };
+        Ok(items)
+    }
+}
 
-        let mut reader: SliceReader = decompressed.borrow();
-        let content_len = reader.len() + HEADER_LEN; // For its header.
+#[derive(Debug, PartialEq)]
+pub struct EntriesBatch<M: MessageExt>(pub Vec<Entries<M>>);
 
-        let mut items_count = codec::decode_var_u64(&mut reader)? as usize;
-        assert!(items_count > 0 && !reader.is_empty());
-        let mut log_batch = LogBatch::with_capacity(items_count);
-        let mut entries_size = 0;
-        while items_count > 0 {
-            let content_offset = (content_len - reader.len()) as u64;
-            let item = LogItem::from_bytes(
-                &mut reader,
-                file_id,
-                base_offset,
-                content_offset,
-                &mut entries_size,
-            )?;
-            log_batch.items_size += item.compute_size();
-            log_batch.items.push(item);
-            items_count -= 1;
-        }
-        assert!(reader.is_empty());
-        buf.consume(batch_len);
-
-        for item in log_batch.items.iter_mut() {
-            if let LogItemContent::Entries(entries) = &mut item.content {
-                entries.update_compression_type(batch_type, batch_len as u64);
-            }
+impl<M> EntriesBatch<M>
+where
+    M: MessageExt,
+{
+    pub fn with_capacity(cap: usize) -> Self {
+        Self(Vec::with_capacity(cap))
+    }
+
+    pub fn merge(&mut self, rhs: &mut Self) {
+        self.0.append(&mut rhs.0);
+    }
+
+    pub fn push(&mut self, entries: Vec<M::Entry>) {
+        self.0.push(Entries::new(entries));
+    }
+
+    pub fn encode_to(
+        &mut self,
+        buf: &mut Vec<u8>,
+        entries_size: &mut usize,
+    ) -> Result<Vec<EntriesIndex>> {
+        let mut entries_indexes = Vec::with_capacity(self.0.len());
+        for entries in self.0.drain(..) {
+            let entries_index = entries.encode_to(buf, entries_size)?;
+            entries_indexes.push(entries_index);
         }
+        Ok(entries_indexes)
+    }
```

Comment:
Maybe this wrapper can be removed too? Doesn't seem very necessary.

Reply:
Do you mean `EntriesBatch` of just these 3 methods?

Reply:
I mean remove `EntriesBatch` and directly store `Vec` in `LogBatch`.